File: 2021/rust/day1.rs
use std::io::{self, BufRead, BufReader};

fn count_greater(iter: impl IntoIterator<Item = u32>) -> usize {
    let mut count = 0;
    let mut last = None;
    for x in iter {
        if let Some(last) = last {
            if last < x {
                count += 1;
            }
        }
        last = Some(x);
    }
    count
}

fn sum_of_triplets<'a>(items: &'a [u32]) -> impl Iterator<Item = u32> + 'a {
    (0..items.len().saturating_sub(2)).map(move |i| {
        items[i] + items[i + 1] + items[i + 2]
    })
}

fn main() -> io::Result<()> {
    let lines: Vec<u32> = std::fs::File::open("inputs/day1.txt")
        .map(BufReader::new)?
        .lines()
        .map(|line| line.map(|x| x.parse().unwrap())).collect::<Result<_, _>>()?;

    println!("Part 1:");
    println!("{}", count_greater(lines.iter().cloned()));
    println!("Part 2:");
    println!("{}", count_greater(sum_of_triplets(&lines)));
    Ok(())
}

File: 2021/python/day1.py
from typing import Iterable, List


def count_times_greater(iter):
    last = None
    count = 0
    for current in iter:
        if last is not None and last < current:
            count += 1
        last = current
    return count


def make_three_windows(items):
    for i in range(len(items) - 2):
        yield items[i], items[i + 1], items[i + 2]


def part1(contents):
    return count_times_greater(contents)


def part2(contents):
    return count_times_greater(map(sum, make_three_windows(contents)))


if __name__ == "__main__":
    with open("inputs/day1.txt") as file:
        contents = [*map(int, file)]

    print("Part 1: ")
    print(part1(contents))
    print("Part 2: ")
    print(part2(contents))
